File: gopher-mcp/sdk/java/gopher-mcp/src/main/java/com/gopher/mcp/filter/McpFilterChain.java
package com.gopher.mcp.filter;

import com.gopher.mcp.filter.type.ProtocolMetadata;
import com.gopher.mcp.filter.type.buffer.FilterCondition;
import com.gopher.mcp.filter.type.chain.*;
import com.gopher.mcp.jna.McpFilterChainLibrary;
import com.gopher.mcp.jna.type.filter.McpProtocolMetadata;
import com.gopher.mcp.jna.type.filter.chain.*;
import com.sun.jna.Native;
import com.sun.jna.NativeLong;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.LongByReference;
import com.sun.jna.ptr.PointerByReference;

/**
 * Java wrapper for the MCP Filter Chain API. Provides one-to-one method mapping to
 * McpFilterChainLibrary.
 *
 * <p>This wrapper provides comprehensive filter chain composition and management, including dynamic
 * routing, conditional execution, and performance optimization.
 */
public class McpFilterChain implements AutoCloseable {

  private final McpFilterChainLibrary lib;
  private Long primaryChainHandle;

  public McpFilterChain() {
    this.lib = McpFilterChainLibrary.INSTANCE;
  }

  // ============================================================================
  // Advanced Chain Builder Methods (one-to-one mapping)
  // ============================================================================

  /** Create chain builder with configuration Maps to: mcp_chain_builder_create_ex */
  public long chainBuilderCreateEx(long dispatcher, ChainConfig config) {
    Pointer dispatcherPtr = dispatcher != 0 ? new Pointer(dispatcher) : null;
    McpChainConfig.ByReference nativeConfig = new McpChainConfig.ByReference();
    nativeConfig.name = config.getName();
    nativeConfig.mode = config.getMode();
    nativeConfig.routing = config.getRouting();
    nativeConfig.max_parallel = config.getMaxParallel();
    nativeConfig.buffer_size = config.getBufferSize();
    nativeConfig.timeout_ms = config.getTimeoutMs();
    nativeConfig.stop_on_error = (byte) (config.isStopOnError() ? 1 : 0);

    Pointer builder = lib.mcp_chain_builder_create_ex(dispatcherPtr, nativeConfig);
    return Pointer.nativeValue(builder);
  }

  /** Add filter node to chain Maps to: mcp_chain_builder_add_node */
  public int chainBuilderAddNode(long builder, FilterNode node) {
    McpFilterNode.ByReference nativeNode = new McpFilterNode.ByReference();
    nativeNode.filter = node.getFilterHandle() != 0 ? new Pointer(node.getFilterHandle()) : null;
    nativeNode.name = node.getName();
    nativeNode.priority = node.getPriority();
    nativeNode.enabled = (byte) (node.isEnabled() ? 1 : 0);
    nativeNode.bypass_on_error = (byte) (node.isBypassOnError() ? 1 : 0);
    nativeNode.config = node.getConfigHandle() != 0 ? new Pointer(node.getConfigHandle()) : null;

    return lib.mcp_chain_builder_add_node(new Pointer(builder), nativeNode);
  }

  /** Add conditional filter Maps to: mcp_chain_builder_add_conditional */
  public int chainBuilderAddConditional(long builder, FilterCondition condition, long filter) {
    McpFilterCondition.ByReference nativeCondition = new McpFilterCondition.ByReference();
    nativeCondition.match_type = condition.getMatchType();
    nativeCondition.field = condition.getField();
    nativeCondition.value = condition.getValue();
    nativeCondition.target_filter =
        condition.getTargetFilter() != 0 ? new Pointer(condition.getTargetFilter()) : null;

    return lib.mcp_chain_builder_add_conditional(new Pointer(builder), nativeCondition, filter);
  }

  /** Add parallel filter group Maps to: mcp_chain_builder_add_parallel_group */
  public int chainBuilderAddParallelGroup(long builder, long[] filters) {
    return lib.mcp_chain_builder_add_parallel_group(
        new Pointer(builder), filters, new NativeLong(filters.length));
  }

  /** Set custom routing function Maps to: mcp_chain_builder_set_router */
  public int chainBuilderSetRouter(
      long builder, McpFilterChainLibrary.MCP_ROUTING_FUNCTION_T router, long userData) {
    Pointer userDataPtr = userData != 0 ? new Pointer(userData) : null;
    return lib.mcp_chain_builder_set_router(new Pointer(builder), router, userDataPtr);
  }

  // ============================================================================
  // Chain Management Methods (one-to-one mapping)
  // ============================================================================

  /** Get chain state Maps to: mcp_chain_get_state */
  public ChainState chainGetState(long chain) {
    int state = lib.mcp_chain_get_state(chain);
    return ChainState.fromValue(state);
  }

  /** Pause chain execution Maps to: mcp_chain_pause */
  public int chainPause(long chain) {
    return lib.mcp_chain_pause(chain);
  }

  /** Resume chain execution Maps to: mcp_chain_resume */
  public int chainResume(long chain) {
    return lib.mcp_chain_resume(chain);
  }

  /** Reset chain to initial state Maps to: mcp_chain_reset */
  public int chainReset(long chain) {
    return lib.mcp_chain_reset(chain);
  }

  /** Enable/disable filter in chain Maps to: mcp_chain_set_filter_enabled */
  public int chainSetFilterEnabled(long chain, String filterName, boolean enabled) {
    return lib.mcp_chain_set_filter_enabled(chain, filterName, (byte) (enabled ? 1 : 0));
  }

  /** Get chain statistics Maps to: mcp_chain_get_stats */
  public int chainGetStats(long chain, ChainStats stats) {
    McpChainStats.ByReference nativeStats = new McpChainStats.ByReference();
    int result = lib.mcp_chain_get_stats(chain, nativeStats);

    if (result == 0) {
      stats.setTotalProcessed(nativeStats.total_processed);
      stats.setTotalErrors(nativeStats.total_errors);
      stats.setTotalBypassed(nativeStats.total_bypassed);
      stats.setAvgLatencyMs(nativeStats.avg_latency_ms);
      stats.setMaxLatencyMs(nativeStats.max_latency_ms);
      stats.setThroughputMbps(nativeStats.throughput_mbps);
      stats.setActiveFilters(nativeStats.active_filters);
    }

    return result;
  }

  /** Set chain event callback Maps to: mcp_chain_set_event_callback */
  public int chainSetEventCallback(
      long chain, McpFilterChainLibrary.MCP_CHAIN_EVENT_CB callback, long userData) {
    Pointer userDataPtr = userData != 0 ? new Pointer(userData) : null;
    return lib.mcp_chain_set_event_callback(chain, callback, userDataPtr);
  }

  // ============================================================================
  // Dynamic Chain Composition Methods (one-to-one mapping)
  // ============================================================================

  /** Create dynamic chain from JSON configuration Maps to: mcp_chain_create_from_json */
  public long chainCreateFromJson(long dispatcher, long jsonConfig) {
    Pointer dispatcherPtr = dispatcher != 0 ? new Pointer(dispatcher) : null;
    Pointer jsonPtr = jsonConfig != 0 ? new Pointer(jsonConfig) : null;

    long handle = lib.mcp_chain_create_from_json(dispatcherPtr, jsonPtr);
    if (this.primaryChainHandle == null && handle != 0) {
      this.primaryChainHandle = handle;
    }
    return handle;
  }

  /** Export chain configuration to JSON Maps to: mcp_chain_export_to_json */
  public long chainExportToJson(long chain) {
    Pointer json = lib.mcp_chain_export_to_json(chain);
    return Pointer.nativeValue(json);
  }

  /** Clone a filter chain Maps to: mcp_chain_clone */
  public long chainClone(long chain) {
    return lib.mcp_chain_clone(chain);
  }

  /** Merge two chains Maps to: mcp_chain_merge */
  public long chainMerge(long chain1, long chain2, ChainExecutionMode mode) {
    return lib.mcp_chain_merge(chain1, chain2, mode.getValue());
  }

  // ============================================================================
  // Chain Router Methods (one-to-one mapping)
  // ============================================================================

  /** Create chain router Maps to: mcp_chain_router_create */
  public long chainRouterCreate(RouterConfig config) {
    McpRouterConfig.ByReference nativeConfig = new McpRouterConfig.ByReference();
    nativeConfig.strategy = config.getStrategy();
    nativeConfig.hash_seed = config.getHashSeed();
    nativeConfig.route_table =
        config.getRouteTable() != 0 ? new Pointer(config.getRouteTable()) : null;
    nativeConfig.custom_router_data =
        config.getCustomRouterData() != null ? new Pointer(Native.malloc(8)) : null;

    Pointer router = lib.mcp_chain_router_create(nativeConfig);
    return Pointer.nativeValue(router);
  }

  /** Add route to router Maps to: mcp_chain_router_add_route */
  public int chainRouterAddRoute(
      long router, McpFilterChainLibrary.MCP_FILTER_MATCH_CB condition, long chain) {
    return lib.mcp_chain_router_add_route(new Pointer(router), condition, chain);
  }

  /** Route buffer through appropriate chain Maps to: mcp_chain_router_route */
  public long chainRouterRoute(long router, long buffer, ProtocolMetadata metadata) {
    McpProtocolMetadata.ByReference nativeMeta = null;
    if (metadata != null) {
      nativeMeta = new McpProtocolMetadata.ByReference();
      nativeMeta.layer = metadata.getLayer();
      // Note: Union fields would need proper handling based on layer
    }

    return lib.mcp_chain_router_route(new Pointer(router), buffer, nativeMeta);
  }

  /** Destroy chain router Maps to: mcp_chain_router_destroy */
  public void chainRouterDestroy(long router) {
    lib.mcp_chain_router_destroy(new Pointer(router));
  }

  // ============================================================================
  // Chain Pool Methods (one-to-one mapping)
  // ============================================================================

  /** Create chain pool for load balancing Maps to: mcp_chain_pool_create */
  public long chainPoolCreate(long baseChain, long poolSize, ChainRoutingStrategy strategy) {
    Pointer pool =
        lib.mcp_chain_pool_create(baseChain, new NativeLong(poolSize), strategy.getValue());
    return Pointer.nativeValue(pool);
  }

  /** Get next chain from pool Maps to: mcp_chain_pool_get_next */
  public long chainPoolGetNext(long pool) {
    return lib.mcp_chain_pool_get_next(new Pointer(pool));
  }

  /** Return chain to pool Maps to: mcp_chain_pool_return */
  public void chainPoolReturn(long pool, long chain) {
    lib.mcp_chain_pool_return(new Pointer(pool), chain);
  }

  /** Get pool statistics Maps to: mcp_chain_pool_get_stats */
  public int chainPoolGetStats(
      long pool,
      PointerByReference active,
      PointerByReference idle,
      LongByReference totalProcessed) {
    return lib.mcp_chain_pool_get_stats(new Pointer(pool), active, idle, totalProcessed);
  }

  /** Destroy chain pool Maps to: mcp_chain_pool_destroy */
  public void chainPoolDestroy(long pool) {
    lib.mcp_chain_pool_destroy(new Pointer(pool));
  }

  // ============================================================================
  // Chain Optimization Methods (one-to-one mapping)
  // ============================================================================

  /** Optimize chain by removing redundant filters Maps to: mcp_chain_optimize */
  public int chainOptimize(long chain) {
    return lib.mcp_chain_optimize(chain);
  }

  /** Reorder filters for optimal performance Maps to: mcp_chain_reorder_filters */
  public int chainReorderFilters(long chain) {
    return lib.mcp_chain_reorder_filters(chain);
  }

  /** Profile chain performance Maps to: mcp_chain_profile */
  public int chainProfile(long chain, long testBuffer, long iterations, PointerByReference report) {
    return lib.mcp_chain_profile(chain, testBuffer, new NativeLong(iterations), report);
  }

  // ============================================================================
  // Chain Debugging Methods (one-to-one mapping)
  // ============================================================================

  /** Enable chain tracing Maps to: mcp_chain_set_trace_level */
  public int chainSetTraceLevel(long chain, int traceLevel) {
    return lib.mcp_chain_set_trace_level(chain, traceLevel);
  }

  /** Dump chain structure Maps to: mcp_chain_dump */
  public String chainDump(long chain, String format) {
    return lib.mcp_chain_dump(chain, format);
  }

  /** Validate chain configuration Maps to: mcp_chain_validate */
  public int chainValidate(long chain, PointerByReference errors) {
    return lib.mcp_chain_validate(chain, errors);
  }

  // ============================================================================
  // AutoCloseable Implementation
  // ============================================================================

  @Override
  public void close() {
    // Chain handles are typically managed through reference counting
    // No explicit destroy needed
    if (primaryChainHandle != null) {
      primaryChainHandle = null;
    }
  }

  // ============================================================================
  // Utility Methods
  // ============================================================================

  /** Get the primary chain handle */
  public Long getPrimaryChainHandle() {
    return primaryChainHandle;
  }

  /** Set the primary chain handle */
  public void setPrimaryChainHandle(long handle) {
    this.primaryChainHandle = handle;
  }
}

File: gopher-mcp/sdk/typescript/src/mcp-filter-chain.ts
/**
 * @file filter-chain.ts
 * @brief TypeScript wrapper for MCP C Filter Chain API (mcp_c_filter_chain.h)
 *
 * This module provides TypeScript wrappers for advanced filter chain management
 * including dynamic composition, conditional execution, parallel processing,
 * routing, and optimization. It uses the existing C++ RAII system through FFI calls.
 */

import * as koffi from "koffi";
import { mcpFilterLib } from "./mcp-ffi-bindings";
import { ValidationResultStruct, AssemblyResultStruct } from "./mcp-c-structs";
import { ProtocolMetadata } from "./mcp-filter-api";

// ============================================================================
// Chain Types and Enumerations (matching mcp_c_filter_chain.h)
// ============================================================================

export enum ChainExecutionMode {
  SEQUENTIAL = 0, // Execute filters in order
  PARALLEL = 1, // Execute filters in parallel
  CONDITIONAL = 2, // Execute based on conditions
  PIPELINE = 3, // Pipeline mode with buffering
}

export enum RoutingStrategy {
  ROUND_ROBIN = 0, // Round-robin distribution
  LEAST_LOADED = 1, // Route to least loaded filter
  HASH_BASED = 2, // Hash-based routing
  PRIORITY = 3, // Priority-based routing
  CUSTOM = 99, // Custom routing function
}

export enum MatchCondition {
  ALL = 0, // Match all conditions
  ANY = 1, // Match any condition
  NONE = 2, // Match no conditions
  CUSTOM = 99, // Custom match function
}

const MCP_OK = 0;

type JsonHandle = any;

export enum ChainState {
  IDLE = 0,
  PROCESSING = 1,
  PAUSED = 2,
  ERROR = 3,
  COMPLETED = 4,
}

// ============================================================================
// Data Structures (matching mcp_c_filter_chain.h)
// ============================================================================

export interface FilterNode {
  filter: number;
  name: string;
  priority: number;
  enabled: boolean;
  bypassOnError: boolean;
  config: any; // JSON configuration
}

export interface ChainConfig {
  name: string;
  mode: ChainExecutionMode;
  routing: RoutingStrategy;
  maxParallel: number;
  bufferSize: number;
  timeoutMs: number;
  stopOnError: boolean;
}

export interface FilterCondition {
  matchType: MatchCondition;
  field: string;
  value: string;
  targetFilter: number;
}

export interface ChainStats {
  totalProcessed: number;
  totalErrors: number;
  totalBypassed: number;
  avgLatencyMs: number;
  maxLatencyMs: number;
  throughputMbps: number;
  activeFilters: number;
}

export interface RouterConfig {
  strategy: RoutingStrategy;
  hashSeed: number;
  routeTable: Record<string, any>; // Map of conditions to chains
  customRouterData: any;
}

// ============================================================================
// Callback Types (matching mcp_c_filter_chain.h)
// ============================================================================

export type RoutingFunction = (
  buffer: number,
  nodes: FilterNode[],
  nodeCount: number,
  userData: any
) => number;

export type ChainEventCallback = (
  chain: number,
  oldState: ChainState,
  newState: ChainState,
  userData: any
) => void;

export type FilterMatchCallback = (
  buffer: number,
  metadata: ProtocolMetadata,
  userData: any
) => boolean;

// ============================================================================
// Assembler-based Chain Configuration
// ============================================================================

export interface FilterConfig {
  type: string;
  name?: string;
  config?: any; // JSON configuration
  enabled?: boolean;
  enabledWhen?: any; // JSON condition
}

export interface FilterChainConfig {
  name?: string;
  transportType?: string;
  filters: FilterConfig[];
}

// ============================================================================
// Canonical Listener-based Configuration (matching C++ ListenerConfig)
// ============================================================================

export interface SocketAddress {
  address: string;
  port_value: number;
}

export interface Address {
  socket_address: SocketAddress;
}

export interface FilterSpec {
  name: string;
  type: string;
  config?: any; // Optional filter-specific configuration
}

export interface FilterChain {
  filters: FilterSpec[];
  name?: string;
  transport_socket?: any; // Optional TLS configuration
}

export interface ListenerConfig {
  name: string;
  address?: Address; // Optional for hybrid SDK mode where SDK handles transport
  filter_chains: FilterChain[];
  per_connection_buffer_limit_bytes?: number;
  metadata?: any;
}

export interface CanonicalConfig {
  listeners: ListenerConfig[];
}

export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

export interface AssemblyResult {
  success: boolean;
  chain?: number;
  errorMessage?: string;
  createdFilters: string[];
  warnings: string[];
}

function normalizeFilterChainConfig(config: FilterChainConfig): Record<string, unknown> {
  const filters = config.filters || [];
  return {
    name: config.name ?? "default",
    transport_type: config.transportType ?? "tcp",
    filters: filters.map((filter, index) => normalizeFilterSpec(filter, index)),
  };
}

function normalizeFilterSpec(filter: FilterConfig, index: number): Record<string, unknown> {
  return {
    type: filter.type,
    name: filter.name ?? `${filter.type || "filter"}:${index}`,
    config: normalizeFilterConfigPayload(filter.type, filter.config ?? {}),
    // Note: enabled and enabled_when fields removed - they're not supported by C API during chain creation
    // These can be set later via mcp_chain_set_filter_enabled() if needed
  };
}

function normalizeFilterConfigPayload(type: string, config: any): Record<string, unknown> {
  if (!config || typeof config !== "object") {
    return config ?? {};
  }

  if (isCircuitBreakerFilter(type)) {
    return normalizeCircuitBreakerConfig(config);
  }

  return config;
}

function isCircuitBreakerFilter(type: string): boolean {
  return type === "circuit_breaker" || type.endsWith(".circuit_breaker");
}

const CIRCUIT_BREAKER_KEY_MAP: Record<string, string> = {
  failureThreshold: "failure_threshold",
  errorRateThreshold: "error_rate_threshold",
  minRequests: "min_requests",
  timeoutMs: "timeout_ms",
  windowSizeMs: "window_size_ms",
  halfOpenMaxRequests: "half_open_max_requests",
  halfOpenSuccessThreshold: "half_open_success_threshold",
  trackTimeouts: "track_timeouts",
  trackErrors: "track_errors",
  track4xxAsErrors: "track_4xx_as_errors",
};

function normalizeCircuitBreakerConfig(config: Record<string, unknown>): Record<string, unknown> {
  const normalized: Record<string, unknown> = { ...config };

  for (const [camel, snake] of Object.entries(CIRCUIT_BREAKER_KEY_MAP)) {
    if (Object.prototype.hasOwnProperty.call(config, camel)) {
      if (normalized[snake] === undefined) {
        normalized[snake] = (config as any)[camel];
      }
      delete normalized[camel];
    }
  }

  return normalized;
}

/**
 * Convert canonical listener configuration to assembler format
 */
function canonicalToAssemblerConfig(config: CanonicalConfig): FilterChainConfig {
  // For now, we'll take the first listener's first filter chain
  // In a full implementation, this could be more sophisticated
  if (!config.listeners || config.listeners.length === 0) {
    throw new Error("No listeners defined in canonical configuration");
  }

  const listener = config.listeners[0];
  if (!listener || !listener.filter_chains || listener.filter_chains.length === 0) {
    throw new Error("No filter chains defined in listener");
  }

  const filterChain = listener.filter_chains[0];
  if (!filterChain) {
    throw new Error("No filter chain at index 0");
  }

  return {
    name: filterChain.name || listener.name,
    transportType: "tcp", // Default, could be inferred from listener type
    filters: (filterChain.filters || []).map(filter => ({
      type: filter.type,
      name: filter.name,
      config: filter.config,
      enabled: true,
    })),
  };
}

/**
 * Check if configuration is in canonical format
 */
function isCanonicalConfig(config: any): config is CanonicalConfig {
  return (
    config && typeof config === "object" && "listeners" in config && Array.isArray(config.listeners)
  );
}

function createJsonHandleFromConfig(config: FilterChainConfig): JsonHandle {
  const normalized = normalizeFilterChainConfig(config);
  const jsonStr = JSON.stringify(normalized);
  const handle = mcpFilterLib.mcp_json_parse(jsonStr);
  if (!handle) {
    throw new Error("Failed to parse filter chain configuration JSON");
  }
  return handle;
}

/**
 * Validate filter chain configuration using the assembler-aware API
 */
export function validateFilterChainConfig(config: FilterChainConfig): ValidationResult {
  const jsonHandle = createJsonHandleFromConfig(config);
  const resultPtr = koffi.alloc(ValidationResultStruct, 1);

  try {
    const status = mcpFilterLib.mcp_chain_validate_json(
      jsonHandle,
      koffi.as(resultPtr, "void*")
    ) as number;

    if (status !== MCP_OK) {
      return { valid: false, errors: [`Validation failed with code: ${status}`], warnings: [] };
    }

    const buffer: Buffer = resultPtr;
    const valid = buffer.readUInt8(0) !== 0;

    return { valid, errors: [], warnings: [] };
  } finally {
    mcpFilterLib.mcp_chain_validation_result_free(koffi.as(resultPtr, "void*"));
    koffi.free(resultPtr);
    mcpFilterLib.mcp_json_free(jsonHandle);
  }
}

/**
 * Assemble filter chain from configuration
 */
export function assembleFilterChain(dispatcher: any, config: FilterChainConfig): AssemblyResult {
  const jsonHandle = createJsonHandleFromConfig(config);
  const resultPtr = koffi.alloc(AssemblyResultStruct, 1);

  try {
    const status = mcpFilterLib.mcp_chain_assemble_from_json(
      dispatcher,
      jsonHandle,
      koffi.as(resultPtr, "void*")
    ) as number;

    const buffer: Buffer = resultPtr;
    const successFlag = buffer.readUInt8(0) !== 0;
    const chainHandle = Number(buffer.readBigUInt64LE(8));

    const success = status === MCP_OK && successFlag;
    let errorMessage: string | undefined;

    if (!success) {
      errorMessage =
        status !== MCP_OK ? `Assembly failed with code: ${status}` : "Assembler reported failure";
    }

    const assemblyResult: AssemblyResult = {
      success,
      createdFilters: [],
      warnings: [],
    };

    if (success && chainHandle !== 0) {
      assemblyResult.chain = chainHandle;
    }

    if (errorMessage) {
      assemblyResult.errorMessage = errorMessage;
    }

    return assemblyResult;
  } finally {
    mcpFilterLib.mcp_chain_assembly_result_free(koffi.as(resultPtr, "void*"));
    koffi.free(resultPtr);
    mcpFilterLib.mcp_json_free(jsonHandle);
  }
}

/**
 * Create filter chain directly from canonical configuration using assembler
 */
export function canonicalConfigToNormalizedJson(config: CanonicalConfig): string {
  const chainConfig = canonicalToAssemblerConfig(config);
  const normalized = normalizeFilterChainConfig(chainConfig);
  return JSON.stringify(normalized);
}

export function createFilterChainFromConfig(
  dispatcher: any, // mcp_dispatcher_t (pointer type)
  config: CanonicalConfig
): number {
  if (!isCanonicalConfig(config)) {
    throw new Error("Configuration must be in canonical format with 'listeners' array");
  }

  const jsonStr = canonicalConfigToNormalizedJson(config);
  const jsonHandle = mcpFilterLib.mcp_json_parse(jsonStr);
  try {
    const rawHandle = mcpFilterLib.mcp_chain_create_from_json(dispatcher, jsonHandle);
    // Convert koffi pointer/uint64_t to number
    const handle = typeof rawHandle === "number" ? rawHandle : Number(rawHandle);
    if (!handle) {
      throw new Error("Failed to create filter chain from configuration");
    }
    return handle;
  } finally {
    mcpFilterLib.mcp_json_free(jsonHandle);
  }
}

// ============================================================================
// Chain Management
// ============================================================================

/**
 * Get chain state
 */
export function getChainState(chain: number): ChainState {
  return mcpFilterLib.mcp_chain_get_state(chain) as ChainState;
}

/**
 * Pause chain execution
 */
export function pauseChain(chain: number): number {
  return mcpFilterLib.mcp_chain_pause(chain) as number;
}

/**
 * Resume chain execution
 */
export function resumeChain(chain: number): number {
  return mcpFilterLib.mcp_chain_resume(chain) as number;
}

/**
 * Reset chain to initial state
 */
export function resetChain(chain: number): number {
  return mcpFilterLib.mcp_chain_reset(chain) as number;
}

/**
 * Enable/disable filter in chain
 */
export function setFilterEnabled(chain: number, filterName: string, enabled: boolean): number {
  return mcpFilterLib.mcp_chain_set_filter_enabled(chain, filterName, enabled) as number;
}

/**
 * Get chain statistics
 */
export function getChainStats(chain: number, stats: ChainStats): number {
  return mcpFilterLib.mcp_chain_get_stats(chain, stats) as number;
}

/**
 * Set chain event callback
 */
export function setChainEventCallback(
  chain: number,
  callback: ChainEventCallback,
  userData: any
): number {
  return mcpFilterLib.mcp_chain_set_event_callback(chain, callback, userData) as number;
}

// ============================================================================
// Dynamic Chain Composition
// ============================================================================

/**
 * Create dynamic chain from JSON configuration
 */
export function createChainFromJson(dispatcher: number, jsonConfig: any): number {
  return mcpFilterLib.mcp_chain_create_from_json(dispatcher, jsonConfig) as number;
}

/**
 * Export chain configuration to JSON
 */
export function exportChainToJson(chain: number): any {
  return mcpFilterLib.mcp_chain_export_to_json(chain);
}

/**
 * Clone a filter chain
 */
export function cloneChain(chain: number): number {
  return mcpFilterLib.mcp_chain_clone(chain) as number;
}

/**
 * Merge two chains
 */
export function mergeChains(chain1: number, chain2: number, mode: ChainExecutionMode): number {
  return mcpFilterLib.mcp_chain_merge(chain1, chain2, mode) as number;
}

// ============================================================================
// Chain Router
// ============================================================================

/**
 * Create chain router
 */
export function createChainRouter(config: RouterConfig): any {
  return mcpFilterLib.mcp_chain_router_create(config);
}

/**
 * Add route to router
 */
export function addRouteToRouter(
  router: any,
  condition: FilterMatchCallback,
  chain: number
): number {
  return mcpFilterLib.mcp_chain_router_add_route(router, condition, chain) as number;
}

/**
 * Route buffer through appropriate chain
 */
export function routeBuffer(router: any, buffer: number, metadata: ProtocolMetadata): number {
  return mcpFilterLib.mcp_chain_router_route(router, buffer, metadata) as number;
}

/**
 * Destroy chain router
 */
export function destroyChainRouter(router: any): void {
  mcpFilterLib.mcp_chain_router_destroy(router);
}

// ============================================================================
// Chain Pool for Load Balancing
// ============================================================================

/**
 * Create chain pool for load balancing
 */
export function createChainPool(
  baseChain: number,
  poolSize: number,
  strategy: RoutingStrategy
): any {
  return mcpFilterLib.mcp_chain_pool_create(baseChain, poolSize, strategy);
}

/**
 * Get next chain from pool
 */
export function getNextChainFromPool(pool: any): number {
  return mcpFilterLib.mcp_chain_pool_get_next(pool) as number;
}

/**
 * Return chain to pool
 */
export function returnChainToPool(pool: any, chain: number): void {
  mcpFilterLib.mcp_chain_pool_return(pool, chain);
}

/**
 * Get pool statistics
 */
export function getChainPoolStats(
  pool: any,
  active: number,
  idle: number,
  totalProcessed: number
): number {
  return mcpFilterLib.mcp_chain_pool_get_stats(pool, active, idle, totalProcessed) as number;
}

/**
 * Destroy chain pool
 */
export function destroyChainPool(pool: any): void {
  mcpFilterLib.mcp_chain_pool_destroy(pool);
}

// ============================================================================
// Chain Optimization
// ============================================================================

/**
 * Optimize chain by removing redundant filters
 */
export function optimizeChain(chain: number): number {
  return mcpFilterLib.mcp_chain_optimize(chain) as number;
}

/**
 * Reorder filters for optimal performance
 */
export function reorderFilters(chain: number): number {
  return mcpFilterLib.mcp_chain_reorder_filters(chain) as number;
}

/**
 * Profile chain performance
 */
export function profileChain(
  chain: number,
  testBuffer: number,
  iterations: number,
  report: any
): number {
  return mcpFilterLib.mcp_chain_profile(chain, testBuffer, iterations, report) as number;
}

// ============================================================================
// Chain Debugging
// ============================================================================

/**
 * Enable chain tracing
 */
export function setChainTraceLevel(chain: number, traceLevel: number): number {
  return mcpFilterLib.mcp_chain_set_trace_level(chain, traceLevel) as number;
}

/**
 * Dump chain structure
 */
export function dumpChain(chain: number, format: string): string {
  const result = mcpFilterLib.mcp_chain_dump(chain, format);
  return result ? result.toString() : "";
}

/**
 * Validate chain configuration
 */
export function validateChain(chain: number, errors: any): number {
  return mcpFilterLib.mcp_chain_validate(chain, errors) as number;
}

// ============================================================================
// Utility Functions for Chain Management
// ============================================================================

/**
 * Create a simple sequential chain using canonical configuration
 */
export function createSimpleChain(
  dispatcher: any, // mcp_dispatcher_t (pointer type)
  filterTypes: string[],
  name: string = "simple-chain",
  port: number = 8080
): number {
  // Build canonical configuration
  const config: CanonicalConfig = {
    listeners: [
      {
        name: `${name}_listener`,
        address: {
          socket_address: {
            address: "127.0.0.1",
            port_value: port,
          },
        },
        filter_chains: [
          {
            filters: filterTypes.map((type, index) => ({
              name: `filter_${index}`,
              type: type,
            })),
          },
        ],
      },
    ],
  };

  return createFilterChainFromConfig(dispatcher, config);
}

/**
 * Create a parallel processing chain
 * Note: Parallel execution is handled by the filter chain implementation
 */
export function createParallelChain(
  dispatcher: any, // mcp_dispatcher_t (pointer type)
  filterTypes: string[],
  maxParallel: number = 4,
  name: string = "parallel-chain",
  port: number = 8080
): number {
  // For parallel chains, we still use the canonical config
  // The parallel execution mode would be specified in the filter config
  const config: CanonicalConfig = {
    listeners: [
      {
        name: `${name}_listener`,
        address: {
          socket_address: {
            address: "127.0.0.1",
            port_value: port,
          },
        },
        filter_chains: [
          {
            filters: filterTypes.map((type, index) => ({
              name: `parallel_filter_${index}`,
              type: type,
              config: {
                execution_mode: "parallel",
                max_parallel: maxParallel,
              },
            })),
          },
        ],
      },
    ],
  };

  return createFilterChainFromConfig(dispatcher, config);
}

/**
 * Create a conditional chain with routing
 * Note: Conditional routing would be implemented via filter configuration
 */
export function createConditionalChain(
  dispatcher: any, // mcp_dispatcher_t (pointer type)
  filterConfigs: Array<{ type: string; condition?: any }>,
  name: string = "conditional-chain",
  port: number = 8080
): number {
  const config: CanonicalConfig = {
    listeners: [
      {
        name: `${name}_listener`,
        address: {
          socket_address: {
            address: "127.0.0.1",
            port_value: port,
          },
        },
        filter_chains: [
          {
            filters: filterConfigs.map((fc, index) => ({
              name: `conditional_filter_${index}`,
              type: fc.type,
              config: fc.condition ? { condition: fc.condition } : undefined,
            })),
          },
        ],
      },
    ],
  };

  return createFilterChainFromConfig(dispatcher, config);
}
